_source/_posts/2021-12-06-spring-data-elasticsearch.md

```diff
@@ -0,0 +1,369 @@
+
+{% img blog/spring-data-elasticsearch/kibana-indexes.png alt:"Kibana Indexes" width:"800" %}{: .center-image }
+
+Choose for example the `tag` index. The **Mappings** tab before persisting any instance will look like the following:
+
+```json
+{
+  "mappings": {
+    "_doc": {
+      "properties": {
+        "_class": {
+          "type": "keyword",
+          "index": false,
+          "doc_values": false
+        }
+      }
+    }
+  }
+}
+```
+
+As you can see, properties are not mapped yet.
+
+Go back to the application, and in the top-right menu go to **Entities** > **Tag** and create some `Tag` entities. Then create some `Blog` and `Post` entities as well. As you can see, a search box is present on the entity list page. Go ahead and test the search functionality. The results page will return the matches with pagination.
+
+{% img blog/spring-data-elasticsearch/entity-search-box.png alt:"Entity Search Box" width:"800" %}{: .center-image }
+
+Go back to Kibana, and display the `Tag` mappings after persistence:
+
+```json
+{
+  "mappings": {
+    "_doc": {
+      "properties": {
+        "_class": {
+          "type": "keyword",
+          "index": false,
+          "doc_values": false
+        },
+        "id": {
+          "type": "text",
+          "fields": {
+            "keyword": {
+              "type": "keyword",
+              "ignore_above": 256
+            }
+          }
+        },
+        "name": {
+          "type": "text",
+          "fields": {
+            "keyword": {
+              "type": "keyword",
+              "ignore_above": 256
+            }
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+Now the `id` and `name` properties have been _dynamically mapped_ with _multi-field mapping_. For example, the `name` field is defined with type `text`, for full-text searches, and it defines a `name.keyword` sub-field for aggregations and sorting:
+
+```json
+"name": {
+  "type": "text",
+  "fields": {
+    "keyword": {
+      "type": "keyword",
+      "ignore_above": 256
+    }
+  }
+}
+```
+
+It is often useful to index the same field in different ways for different purposes. Text fields are searchable by default, but by default are not available for aggregations, sorting, unless multi-field mapping is in place. There is another option for enabling sorting for text fields, which is the [`fielddata` mapping parameter](https://www.elastic.co/guide/en/elasticsearch/reference/current/text.html#fielddata-mapping-param), which will not be covered in this tutorial, as it is discouraged in the Elasticsearch documentation, because of high memory consumption.
```

Comment:
```suggestion
It is often useful to index the same field in different ways for different purposes. Text fields are searchable by default, but by default are not available for aggregations, sorting, unless multi-field mapping is in place. There is another option for enabling sorting for text fields, which is the [`fielddata` mapping parameter](https://www.elastic.co/guide/en/elasticsearch/reference/current/text.html#fielddata-mapping-param). This option will not be covered in our tutorial, as it is discouraged in the Elasticsearch documentation, because of high memory consumption.
```
Make two sentences for ease of reading.